From 2cb9bb45a31fbabc4d799011d8a161453c13a708 Mon Sep 17 00:00:00 2001
From: timedout <git@nexy7574.co.uk>
Date: Wed, 29 Jul 2026 17:18:46 +0100
Subject: [PATCH] feat: Implement sketchy backwards compatibility for illegal
 invites

feat: Support accepting invites consisting of solely stripped state

fix: `validate_legacy_invite_state_event` isn't async

fix: Don't panic when m.room.create event isn't a PDU
---
 conduwuit-example.toml       |   9 ++
 src/api/server/invite.rs     | 222 ++++++++++++++++++++++++-----------
 src/api/server/send_join.rs  |   2 +-
 src/api/server/send_knock.rs |   2 +-
 src/api/server/send_leave.rs |   2 +-
 src/api/server/utils.rs      |  35 ++++--
 src/core/config/mod.rs       |   9 ++
 7 files changed, 199 insertions(+), 82 deletions(-)

diff --git a/conduwuit-example.toml b/conduwuit-example.toml
index 24c5159ca..3f26a6430 100644
--- a/conduwuit-example.toml
+++ b/conduwuit-example.toml
@@ -1663,6 +1663,15 @@
 #
 #send_messages_from_ignored_users_to_client = false
 
+# Disables strict validation on incoming federated invites. This weakens
+# your server's security - only enable it if you encounter issues
+# receiving invites from outdated servers.
+#
+# Validation will still be performed, but it will log a warning instead of
+# causing an error.
+#
+#enable_legacy_invite_support = false
+
 # Send "org.matrix.dummy_event" events to the client. This is a debugging
 # option.
 #
diff --git a/src/api/server/invite.rs b/src/api/server/invite.rs
index f9ad2dbb6..6b4eb57bc 100644
--- a/src/api/server/invite.rs
+++ b/src/api/server/invite.rs
@@ -3,7 +3,7 @@
 use axum::extract::State;
 use base64::{Engine as _, engine::general_purpose};
 use conduwuit::{
-	Err, Error, EventTypeExt, PduEvent, Result, debug, err, error,
+	Err, Error, EventTypeExt, PduEvent, Result, debug, debug_warn, err, error,
 	matrix::{Event, StateKey},
 	result::FlatOk,
 	state_res, trace,
@@ -11,8 +11,8 @@
 	warn,
 };
 use ruma::{
-	CanonicalJsonObject, CanonicalJsonValue, OwnedEventId, OwnedRoomId, OwnedUserId, ServerName,
-	UserId,
+	CanonicalJsonObject, CanonicalJsonValue, EventId, OwnedEventId, OwnedRoomId, OwnedUserId,
+	ServerName, UserId,
 	api::{
 		error::{ErrorKind, IncompatibleRoomVersionErrorData},
 		federation::membership::{RawStrippedState, create_invite},
@@ -21,6 +21,7 @@
 	room_version_rules::RoomVersionRules,
 };
 use serde::Deserialize;
+use serde_json::value::RawValue;
 
 use crate::{Ruma, server::utils::validate_any_membership_event};
 
@@ -168,10 +169,14 @@ pub(crate) async fn create_invite_route(
 		// the invite request, but we will get that during send_join anyway.
 		// This is safe to just add directly as an outlier as we already auth
 		// checked it during validation.
-		services
-			.rooms
-			.outlier
-			.add_pdu_outlier(&create_event_id, create_event_json);
+		if let Some(create_event_id) = create_event_id {
+			let mut create_event_json = create_event_json.clone();
+			create_event_json.insert("event_id".to_owned(), create_event_id.as_str().into());
+			services
+				.rooms
+				.outlier
+				.add_pdu_outlier(&create_event_id, &create_event_json);
+		}
 
 		services
 			.rooms
@@ -241,10 +246,10 @@ pub(crate) async fn create_invite_route(
 /// [spec]: https://spec.matrix.org/v1.19/server-server-api/#put_matrixfederationv2inviteroomideventid
 async fn validate_invite_membership_event(
 	services: &crate::State,
-	body: &serde_json::value::RawValue,
+	body: &RawValue,
 	room_version_rules: &RoomVersionRules,
 	origin: &ServerName,
-	create_event_id: OwnedEventId,
+	create_event_id: Option<OwnedEventId>,
 	room_id: OwnedRoomId,
 	event_id: OwnedEventId,
 ) -> Result<(CanonicalJsonObject, OwnedUserId, OwnedUserId)> {
@@ -279,6 +284,9 @@ async fn validate_invite_membership_event(
 	Ok((pdu, sender_user, recipient_user))
 }
 
+type ValidatedInviteState =
+	(Option<OwnedEventId>, HashMap<(StateEventType, StateKey), CanonicalJsonObject>);
+
 /// Validates the *invite state* of an invite request, per the steps listed
 /// under the endpoint's [spec].
 ///
@@ -290,7 +298,8 @@ async fn validate_invite_state(
 	invite_state: &[RawStrippedState],
 	room_version_rules: &RoomVersionRules,
 	room_id: OwnedRoomId,
-) -> Result<(OwnedEventId, HashMap<(StateEventType, StateKey), CanonicalJsonObject>)> {
+) -> Result<ValidatedInviteState> {
+	let allow_stripped = services.config.enable_legacy_invite_support;
 	trace!(?invite_state, "Raw invite state");
 	let mut invite_state_map: HashMap<(StateEventType, StateKey), _> =
 		HashMap::with_capacity(invite_state.len());
@@ -299,49 +308,69 @@ async fn validate_invite_state(
 	for (idx, invite_state_event) in invite_state.iter().cloned().enumerate() {
 		trace!(%idx, ?invite_state_event, "Invite state event");
 		// Stripped state hasn't been sent over federation since v1.16.
-		let RawStrippedState::Pdu(raw_pdu) = invite_state_event else {
-			debug!(%idx, "Invite state event is not a PDU");
-			return Err!(Request(InvalidParam(
-				"PDU in invite state (index {idx}) violates the room event format"
-			)));
+		// we allow stripped state for compatibility with outdated servers if enabled.
+		#[allow(deprecated)]
+		let (raw_pdu, full) = match invite_state_event {
+			| RawStrippedState::Pdu(pdu) => (pdu, true),
+			| RawStrippedState::Stripped(event) if allow_stripped => {
+				warn!(index=%idx, "Event in incoming invite state is not a PDU and cannot be verified");
+				(
+					serde_json::value::to_raw_value(&event)
+						.expect("must be able to convert raw stripped state event to JSON value"),
+					false,
+				)
+			},
+			| RawStrippedState::Stripped(_) => {
+				debug_warn!(%idx, "Invite state event is not a PDU");
+				return Err!(Request(InvalidParam(
+					"PDU in invite state (index {idx}) violates the room event format"
+				)));
+			},
+			| _ =>
+				return Err!(Request(BadJson(
+					"PDU in invite state (index {idx}) is completely malformed"
+				))),
+		};
+		let (state_event_id, state_event_json) = if !allow_stripped || full {
+			validate_invite_state_pdu(services, &raw_pdu, &room_id, room_version_rules).await?
+		} else {
+			validate_legacy_invite_state_event(&raw_pdu, idx)?
 		};
-		let (state_event_room_id, state_event_id, state_event_json) = services
-			.rooms
-			.event_handler
-			.parse_incoming_pdu(&raw_pdu, Some(room_version_rules))
-			.await
-			.map_err(|e| {
-				err!(Request(InvalidParam(debug_warn!("Invalid PDU in invite state: {e}"))))
-			})?;
-
-		if state_event_room_id != room_id {
-			return Err!(Request(InvalidParam(debug_warn!(
-				%state_event_room_id,
-				%room_id,
-				"PDU in invite state ({state_event_id}) belongs to the wrong room"
-			))));
-		}
-
-		services
-			.server_keys
-			.verify_event_json(&state_event_json, room_version_rules)
-			.await
-			.map_err(|e| {
-				err!(Request(InvalidParam("Signature verification failed on invite event: {e}")))
-			})?;
 
 		let Some(state_key) = state_event_json.get("state_key").and_then(|k| k.as_str()) else {
 			return Err!(Request(InvalidParam(debug_info!(
-				"PDU in invite state ({state_event_id}) is not a state event"
+				"Event in invite state ({state_event_id}) is not a state event"
 			))));
 		};
 		let Some(event_type) = state_event_json.get("type").and_then(|k| k.as_str()) else {
 			return Err!(Request(InvalidParam(debug_warn!(
-				"PDU in invite state ({state_event_id}) is not an event?"
+				"Event in invite state ({state_event_id}) is not an event?"
 			))));
 		};
 
 		let key = StateEventType::from(event_type).with_state_key(state_key);
+		if key.0 == StateEventType::RoomCreate && key.1.is_empty() {
+			// Ensure this is a legal create event.
+			match PduEvent::from_id_val(&state_event_id, state_event_json.clone()) {
+				| Ok(pdu_event) => {
+					debug!("Validating discovered create event in invite room state");
+					create_event_id = Some(
+						validate_invite_create_event(&pdu_event, room_version_rules)
+							.await
+							.map(|_| state_event_id.clone())?,
+					)
+				},
+				| Err(e) => {
+					if !allow_stripped {
+						return Err!(Request(InvalidParam(
+							"Invalid create event in invite room state: {e:?}"
+						)));
+					}
+					warn!(error=?e, "Invalid create event in invite room state");
+				},
+			};
+		}
+
 		match invite_state_map.entry(key) {
 			| Entry::Occupied(entry) =>
 				return Err!(Request(InvalidParam(
@@ -349,41 +378,98 @@ async fn validate_invite_state(
 					entry.key(),
 				))),
 			| Entry::Vacant(entry) => {
-				if entry.key().0 == StateEventType::RoomCreate {
-					// Ensure this is a legal create event.
-					let pdu_event =
-						PduEvent::from_id_val(&state_event_id, state_event_json.clone())
-							.expect("must be able to create pdu event from event json");
-					debug!("Validating discovered create event in invite room state");
-					validate_invite_create_event(&pdu_event, room_version_rules).await?;
-					create_event_id = Some(state_event_id);
-				}
 				entry.insert(state_event_json);
 			},
 		}
 	}
-	let Some(create_event_id) = create_event_id else {
-		return Err!(Request(InvalidParam(debug_warn!(
-			parsed_state=?invite_state_map,
-			"Invite state does not contain the m.room.create event"
-		))));
-	};
-	invite_state_map.iter().try_for_each(|(key, event_json)| {
-		service::rooms::event_handler::Service::pdu_format_check_1(
-			event_json,
-			room_version_rules,
-			&create_event_id,
-		)
-		.map_err(|e| {
-			err!(Request(InvalidParam(
-				"PDU in invite state for {key:?} violates the room event format: {e}"
-			)))
-		})
-	})?;
+
+	format_check_state_map(
+		create_event_id.as_deref(),
+		&invite_state_map,
+		room_version_rules,
+		allow_stripped,
+	)?;
 
 	Ok((create_event_id, invite_state_map))
 }
 
+async fn validate_invite_state_pdu(
+	services: &crate::State,
+	raw_pdu: &RawValue,
+	room_id: &ruma::RoomId,
+	rules: &RoomVersionRules,
+) -> Result<(OwnedEventId, CanonicalJsonObject)> {
+	let (state_event_room_id, state_event_id, state_event_json) = services
+		.rooms
+		.event_handler
+		.parse_incoming_pdu(raw_pdu, Some(rules))
+		.await
+		.map_err(|e| {
+			err!(Request(InvalidParam(debug_warn!("Invalid PDU in invite state: {e}"))))
+		})?;
+	if state_event_room_id != room_id {
+		return Err!(Request(InvalidParam(debug_warn!(
+			%state_event_room_id,
+			%room_id,
+			"PDU in invite state ({state_event_id}) belongs to the wrong room"
+		))));
+	}
+
+	services
+		.server_keys
+		.verify_event_json(&state_event_json, rules)
+		.await
+		.map_err(|e| {
+			err!(Request(InvalidParam("Signature verification failed on invite event: {e}")))
+		})?;
+	Ok((state_event_id, state_event_json))
+}
+
+fn validate_legacy_invite_state_event(
+	event: &RawValue,
+	idx: usize,
+) -> Result<(OwnedEventId, CanonicalJsonObject)> {
+	let pdu = serde_json::from_str::<CanonicalJsonObject>(event.get()).map_err(|e| {
+		err!(BadServerResponse(debug_warn!("Error parsing incoming event {e:?}")))
+	})?;
+	Ok((EventId::parse(format!("$stripped_invite_state_{idx}"))?, pdu))
+}
+
+fn format_check_state_map(
+	create_event_id: Option<&EventId>,
+	invite_state_map: &HashMap<(StateEventType, StateKey), CanonicalJsonObject>,
+	room_version_rules: &RoomVersionRules,
+	allow_stripped: bool,
+) -> Result<()> {
+	let Some(create_event_id) = create_event_id else {
+		if !allow_stripped {
+			return Err!(Request(InvalidParam(debug_warn!(
+				parsed_state=?invite_state_map,
+				"Invite state does not contain a valid m.room.create event"
+			))));
+		}
+
+		warn!("Not validating final invite state map, no valid create event was included.");
+		return Ok(());
+	};
+	invite_state_map.iter().try_for_each(|(key, event_json)| {
+		if event_json.get("signatures").is_none() && allow_stripped {
+			warn!(state_key=?key, "Skipping validation of nonconformant invite state event");
+			Ok(())
+		} else {
+			service::rooms::event_handler::Service::pdu_format_check_1(
+				event_json,
+				room_version_rules,
+				create_event_id,
+			)
+			.map_err(|e| {
+				err!(Request(InvalidParam(
+					"PDU in invite state for {key:?} violates the room event format: {e}"
+				)))
+			})
+		}
+	})
+}
 #[derive(Deserialize)]
 struct MFederate {
 	#[serde(rename = "m.federate")]
diff --git a/src/api/server/send_join.rs b/src/api/server/send_join.rs
index 7b0e2f2df..aef46473b 100644
--- a/src/api/server/send_join.rs
+++ b/src/api/server/send_join.rs
@@ -252,7 +252,7 @@ pub(crate) async fn create_join_event_v2_route(
 		&services,
 		&body.pdu,
 		&room_version_rules,
-		create_event.event_id.clone(),
+		Some(create_event.event_id.clone()),
 		body.room_id.clone(),
 		body.event_id.clone(),
 	)
diff --git a/src/api/server/send_knock.rs b/src/api/server/send_knock.rs
index 15edc6a11..4edf3941f 100644
--- a/src/api/server/send_knock.rs
+++ b/src/api/server/send_knock.rs
@@ -67,7 +67,7 @@ pub(crate) async fn create_knock_event_v1_route(
 		&services,
 		&body.pdu,
 		&room_version_rules,
-		create_event.event_id().to_owned(),
+		Some(create_event.event_id().to_owned()),
 		body.room_id.clone(),
 		body.event_id.clone(),
 	)
diff --git a/src/api/server/send_leave.rs b/src/api/server/send_leave.rs
index cbbb4d475..e41cdaf12 100644
--- a/src/api/server/send_leave.rs
+++ b/src/api/server/send_leave.rs
@@ -53,7 +53,7 @@ pub(crate) async fn create_leave_event_v2_route(
 		&services,
 		&body.pdu,
 		&room_version_rules,
-		create_event.event_id.clone(),
+		Some(create_event.event_id.clone()),
 		body.room_id.clone(),
 		body.event_id.clone(),
 	)
diff --git a/src/api/server/utils.rs b/src/api/server/utils.rs
index a6803eb61..efe6db547 100644
--- a/src/api/server/utils.rs
+++ b/src/api/server/utils.rs
@@ -80,6 +80,12 @@ pub(super) async fn assert(&self) -> Result {
 /// 6. `state_key` field presence (and parsing)
 /// 7. PDU room format check (PDU check 1)
 ///
+/// ## Compatibility
+///
+/// If `create_event_id` is `None`, step 7 is not performed. This should only be
+/// the case for the invite route where servers still send stripped events on
+/// occasion, and as such the create event may not be included in the state.
+///
 /// ## Returns
 ///
 /// A resulting tuple of (PDU JSON, target membership state, sender, recipient).
@@ -87,7 +93,7 @@ pub(crate) async fn validate_any_membership_event(
 	services: &crate::State,
 	body: &serde_json::value::RawValue,
 	room_version_rules: &RoomVersionRules,
-	create_event_id: OwnedEventId,
+	create_event_id: Option<OwnedEventId>,
 	expected_room_id: OwnedRoomId,
 	expected_event_id: OwnedEventId,
 ) -> Result<(CanonicalJsonObject, MembershipState, OwnedUserId, OwnedUserId)> {
@@ -171,6 +177,11 @@ pub(crate) async fn validate_any_membership_event(
 		.ok_or_else(|| err!(Request(BadJson("Event is not a string"))))?
 		.to_owned();
 
+	assert!(
+		create_event_id.is_some() || membership == "invite",
+		"non-invite membership check had no associated create event"
+	);
+
 	let sender_user = pdu
 		.get("sender")
 		.and_then(|v| v.as_str())
@@ -184,16 +195,18 @@ pub(crate) async fn validate_any_membership_event(
 		.and_then(Result::ok)
 		.ok_or_else(|| err!(Request(InvalidParam("Invalid state_key property"))))?;
 
-	// Do a quick format check. The spec doesn't suggest this, but it's probably
-	// a good idea nonetheless.
-	service::rooms::event_handler::Service::pdu_format_check_1(
-		&pdu,
-		room_version_rules,
-		&create_event_id,
-	)
-	.map_err(|e| {
-		err!(Request(InvalidParam("Membership event violates the room event format: {e}")))
-	})?;
+	if let Some(create_event_id) = create_event_id {
+		// Do a quick format check. The spec doesn't suggest this, but it's probably
+		// a good idea nonetheless.
+		service::rooms::event_handler::Service::pdu_format_check_1(
+			&pdu,
+			room_version_rules,
+			&create_event_id,
+		)
+		.map_err(|e| {
+			err!(Request(InvalidParam("Membership event violates the room event format: {e}")))
+		})?;
+	}
 
 	Ok((pdu, membership.into(), sender_user, recipient_user))
 }
diff --git a/src/core/config/mod.rs b/src/core/config/mod.rs
index 3fc9566ca..b180751c2 100644
--- a/src/core/config/mod.rs
+++ b/src/core/config/mod.rs
@@ -1946,6 +1946,15 @@ pub struct Config {
 	#[serde(default)]
 	pub send_messages_from_ignored_users_to_client: bool,
 
+	/// Disables strict validation on incoming federated invites. This weakens
+	/// your server's security - only enable it if you encounter issues
+	/// receiving invites from outdated servers.
+	///
+	/// Validation will still be performed, but it will log a warning instead of
+	/// causing an error.
+	#[serde(default)]
+	pub enable_legacy_invite_support: bool,
+
 	/// Send "org.matrix.dummy_event" events to the client. This is a debugging
 	/// option.
 	#[serde(default)]
-- 
2.55.0

